oauth.py

```diff
@@ -0,0 +1,101 @@
+import json
+from nose.tools import set_trace
+
+from core.util.problem_detail import ProblemDetail as pd
+from config import Configuration
+from oauth2client import client as GoogleClient
+
+class GoogleAuthService(object):
+
+    def __init__(self, client_json_file, redirect_uri, test_mode=False):
+        if test_mode:
+            self.client = DummyGoogleClient()
+        else:
+            self.client = GoogleClient.flow_from_clientsecrets(
+                client_json_file,
+                scope='https://www.googleapis.com/auth/userinfo.email',
+                redirect_uri=redirect_uri
+            )
+
+    def auth_uri(self, redirect_url):
+        return self.client.step1_get_authorize_url(state=redirect_url)
+
+    @classmethod
+    def from_environment(cls, redirect_uri, test_mode=False):
+        config = Configuration.integration(
+            Configuration.GOOGLE_OAUTH_INTEGRATION
+        )
+        client_json_file = config[Configuration.GOOGLE_OAUTH_CLIENT_JSON]
+        return cls(client_json_file, redirect_uri, test_mode)
+
+    def callback(self, request={}):
+        """Google OAuth sign-in flow"""
+
+        # The Google OAuth client sometimes hits the callback with an error.
+        # These will be returned as a problem detail.
+        error = request.get('error')
+        if error:
+            return self.google_error_problem_detail(error)
+        auth_code = request.get('code')
+        if auth_code:
+            redirect_url = request.get("state")
+            credentials = self.client.step2_exchange(auth_code)
+            return dict(
+                email=credentials.id_token.get('email'),
+                access_token=credentials.get_access_token()[0],
+                credentials=credentials.to_json(),
+            ), redirect_url
+
+    def google_error_problem_detail(self, error):
+        detail = "There was an error connecting with Google OAuth: %s" % error
+        return pd(
+            "http://librarysimplified.org/terms/problem/google-oauth-error",
```

Comment:
This should be in the module with the rest of the problem detail constants, since we occasionally need to provide a list of all the possible problem detail URLs.